File: backend/app/models/accrual_amortization.py
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


ScheduleType = Literal["prepaid_amortization", "accrued_expense", "deferred_revenue", "loan_interest"]
ScheduleStatus = Literal["active", "paused", "completed", "terminated"]


class AccountingScheduleCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    account_set_id: str = Field(default="default", min_length=1, max_length=64)
    schedule_code: str = Field(min_length=1, max_length=80)
    schedule_type: ScheduleType
    start_period: str = Field(pattern=r"^\d{4}-\d{2}$")
    end_period: str = Field(pattern=r"^\d{4}-\d{2}$")
    total_amount: Decimal = Field(gt=Decimal("0"), max_digits=16, decimal_places=2)
    debit_account_code: str = Field(min_length=1, max_length=32)
    credit_account_code: str = Field(min_length=1, max_length=32)
    department_id: str | None = Field(default=None, max_length=64)
    project_id: str | None = Field(default=None, max_length=64)


class AccountingSchedule(AccountingScheduleCreate):
    status: ScheduleStatus = "active"
    posted_periods: list[str] = Field(default_factory=list)


class LoanScheduleCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    account_set_id: str = Field(default="default", min_length=1, max_length=64)
    loan_code: str = Field(min_length=1, max_length=80)
    principal: Decimal = Field(gt=Decimal("0"), max_digits=16, decimal_places=2)
    annual_rate: Decimal = Field(gt=Decimal("0"), max_digits=10, decimal_places=6)
    start_period: str = Field(pattern=r"^\d{4}-\d{2}$")
    end_period: str = Field(pattern=r"^\d{4}-\d{2}$")
    loan_account_code: str = Field(default="2001", min_length=1, max_length=32)
    interest_expense_account_code: str = Field(default="6603", min_length=1, max_length=32)
    interest_payable_account_code: str = Field(default="2231", min_length=1, max_length=32)


class LoanSchedule(LoanScheduleCreate):
    status: ScheduleStatus = "active"
    interest_posted_periods: list[str] = Field(default_factory=list)


class AccountingScheduleListResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    account_set_id: str
    total_schedules: int
    total_loans: int
    schedules: list[AccountingSchedule]
    loan_schedules: list[LoanSchedule]


class SchedulePostRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    account_set_id: str = Field(default="default", min_length=1, max_length=64)
    period: str = Field(pattern=r"^\d{4}-\d{2}$")


class LoanInterestPostRequest(LoanScheduleCreate):
    period: str = Field(pattern=r"^\d{4}-\d{2}$")

File: frontend/src/types/accrualAmortization.ts
import type { JournalEntryRecord } from "./accounting";

export type MoneyValue = number | string;

export type ScheduleType = "prepaid_amortization" | "accrued_expense" | "deferred_revenue" | "loan_interest";
export type ScheduleStatus = "active" | "paused" | "completed" | "terminated";

export interface AccountingScheduleCreateRequest {
  account_set_id: string;
  schedule_code: string;
  schedule_type: ScheduleType;
  start_period: string;
  end_period: string;
  total_amount: MoneyValue;
  debit_account_code: string;
  credit_account_code: string;
  department_id?: string | null;
  project_id?: string | null;
}

export interface AccountingSchedule extends AccountingScheduleCreateRequest {
  status: ScheduleStatus;
  posted_periods: string[];
}

export interface LoanSchedule {
  account_set_id: string;
  loan_code: string;
  principal: MoneyValue;
  annual_rate: MoneyValue;
  start_period: string;
  end_period: string;
  loan_account_code: string;
  interest_expense_account_code: string;
  interest_payable_account_code: string;
  status: ScheduleStatus;
  interest_posted_periods: string[];
}

export interface AccrualAmortizationScheduleListResponse {
  account_set_id: string;
  total_schedules: number;
  total_loans: number;
  schedules: AccountingSchedule[];
  loan_schedules: LoanSchedule[];
}

export interface SchedulePostRequest {
  account_set_id: string;
  period: string;
}

export interface LoanInterestPostRequest {
  account_set_id: string;
  loan_code: string;
  period: string;
  principal: MoneyValue;
  annual_rate: MoneyValue;
  start_period: string;
  end_period: string;
  loan_account_code?: string;
  interest_expense_account_code?: string;
  interest_payable_account_code?: string;
}

export type AccrualAmortizationEntryResponse = JournalEntryRecord;
